fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -470,6 +466,23 @@ private ScoredPlan planFilterWithInJoin(@Nonnull PlanContext planContext, @Nonnu
         return null;
     }
 
+    // Get the key expression for the index entries, which includes primary key fields for normal indexes.
+    private KeyExpression indexKeyExpressionForPlan(@Nullable KeyExpression commonPrimaryKey, @Nonnull Index index) {
+        KeyExpression indexKeyExpression = index.getRootExpression();
+        if (indexKeyExpression instanceof KeyWithValueExpression) {
+            indexKeyExpression = ((KeyWithValueExpression) indexKeyExpression).getKeyExpression();
+        }
+        if (commonPrimaryKey != null && indexTypes.getValueTypes().contains(index.getType())) {
+            final List<KeyExpression> keys = new ArrayList<>(commonPrimaryKey.normalizeKeyForPositions());
```

Comment:
This works as `commonPrimaryKey.createsDuplicates() == false`?

Reply:
Right. Duplicates aren't allowed for a primary key. The validator should complain at definition time and save record does `evaluateSingleton`.

In some sense, what you'd like is for the index maintainer to tell the planner what the physical layout of the index entries is. But there isn't actually a vocabulary for that. It might seem that key expressions are, but, in fact, the maintainer is free to interpret those however it chooses. Their evaluation methods are only a convenience for a conventional interpretation. So, really, I suspect, there needs to be a planner-oriented intermediate form for this. That would also be an opportunity to introduce rigorous correlation.

Using normalized key expressions and only doing this for value indexes is kind of a compromise until that comes along.